### DISCUSSION TITLE
outputStandalone not working with Yarn PnP

### DISCUSSION THREAD

**Discussion by iivankin** [vercel/next.js#34600] [Ideas]
### Verify canary release

- [x] I verified that the issue exists in Next.js canary release

### Provide environment information

Operating System:
    Platform: linux
    Arch: x64
    Version: #1 SMP Fri Apr 2 22:23:49 UTC 2021
Binaries:
    Node: 17.5.0
    npm: 8.4.1
    Yarn: 3.1.1
Relevant packages:
   next: 12.1.1-canary.1
   react: 17.0.2
   react-dom: 17.0.2

### What browser are you using? (if relevant)

-

### How are you deploying your application? (if relevant)

-

### Describe the Bug

When building with `outputStandalone` and Yarn PnP enabled, the dependencies are not copied correctly at all (because pnp has a slightly different file structure).
Only `.yarn/cache` is created in the `standalone` folder, which contains partially (!) unpacked next and yarn.
And so, without node_modules we get this error
```
Error: Cannot find module 'next/dist/server/next-server'
Require stack: /.next/standalone/server.js
```

### Expected Behavior

To copy everything correctly.
This is the only thing that does not work with pnp.  It would be very cool if you could fix it

### To Reproduce

- Clone [reproduction repo](https://github.com/iivankin/next-pnp-reproduction)
- Run `yarn`
- Run `yarn build`
- Go to the `.next/standalone` and verify the error.

**Comment by iivankin** [comment]
I see the same error happened at #33564

**Comment by balazsorban44** [comment]
Hi,

The necessary file tracing is done by the `@vercrl/nft` package which currently does not support PnP if I'm correct.

https://github.com/vercel/nft


So unfortunately, I don't think this would be possible right now.

As such, I converted this to an idea. If it gains traction from the community, we might consider revisiting.

Please add your thoughts/feedback in the comments.

Thanks! 🙏

**Reply by rubenhelsloot** [reply]
I'm running into the same issue with a Yarn PnP monorepo setup, and it's forcing me to run `next start` on the 1.2GB builder image instead of `yarn node server.js` on the 110MB runner image. Is there any chance that this could be picked up soon?

**EDIT**
I've managed to reduce the impact on size from 1.2GB to 660MB by doing removing `./.yarn/cache` and adding the entire cache from the build state instead.

```
COPY --from=builder --chown=nextjs:nodejs /app/.pnp.cjs ./
COPY --from=builder --chown=nextjs:nodejs /app/packages/hq/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/packages/hq/.next/static ./.next/static
COPY --from=builder --chown=nextjs:nodejs /app/packages/hq/public ./public

# Copying the entire yarn cache because of an issue with next / @vercel/nft
# https://github.com/vercel/next.js/discussions/34600
RUN rm -rf ./.yarn/cache
COPY --from=builder --chown=nextjs:nodejs /app/.yarn/./.yarn/
```

**Comment by iivankin** [comment]
After debugging i found that [this](https://github.com/vercel/nft/blob/main/src/resolve-dependency.ts#L268) thing is actually broken.
Is it hard to fix?

**Comment by iivankin** [comment]
@balazsorban44 Well, I was able to make it work by copy-pasting 5 lines of code from [here](https://github.com/vercel/nft/pull/152/commits/b484e2da1fc5a896bf4c018bea357ddcc013b5f8#diff-2eebdbf3ec07cd4b5bb39605dd391ebc4cc1dee03f006179e126de1df3404e6fR179)...
Is there any way to do this without modifying the compiled code?

**Reply by iivankin** [reply]
The `nodeFileTrace` func which is called during the build has [resolve](https://github.com/vercel/nft/blob/cb9fce7e30b04816e74631f0d6da9f731c10f869/src/types.ts#L51) hook in the options.
We need to be able to configure this (maybe next.config.js)

**Comment by zigang93** [comment]
@iivankin do u found any hotfix solution ??

**Comment by iivankin** [comment]
@zigang93 
Install [`next-pnp-resolve`](https://www.npmjs.com/package/next-pnp-resolve) package and follow instructions in the README.

I will create pull request later.

**Reply by iivankin** [reply]
Here we go https://github.com/vercel/next.js/pull/34831

**Reply by rubenhelsloot** [reply]
@ijjk closed the pull request, saying "discussion can continue on the linked issue", but I don't see the discussion continued anywhere.

I know it's been 9 months, but what would be necessary to determine a possible approach? I would be open to contributing

**Reply by nalejandroveron** [reply]
@ijjk What are the next steps for this? You closed a valid PR for concerns but haven't got any way to actually solve this.

**Comment by nielsvanmidden** [comment]
Any news on this topic?
